rust/src/jni_utils.rs

```diff
@@ -440,6 +468,48 @@ pub(crate) unsafe fn f64_from_jobject(obj: jobject, jni_env: *mut JNIEnv) -> err
     }
 }
 
+macro_rules! primitive_array_from_jobject {
+    ($fn_name:ident, $rust_type:ty, $get_array_element:path, $release_array_element:path) => {
+        pub(crate) unsafe fn $fn_name(obj: jobject, jni_env: *mut JNIEnv) -> errors::Result<Vec<$rust_type>> {
+            if obj.is_null() {
+                Err(errors::J4RsError::JniError(
+                    format!("Attempt to create an {} array from null", stringify!($rust_type)),
+                ))
+            } else {
+                let length = (opt_to_res(cache::get_jni_get_array_length())?)(
+                    jni_env,
+                    obj
+                );
+                let bytes = (opt_to_res($get_array_element())?)(
+                    jni_env,
+                    obj,
+                    ptr::null_mut()
+                );
+                if bytes.is_null() { return Err(errors::J4RsError::JniError(format!("{} failed", stringify!($get_array_element)))) }
+                let parts_mut = std::slice::from_raw_parts_mut(bytes as *mut $rust_type, length as usize);
```

Comment:
What if length is zero? According to the rust docs for [from_raw_parts_mut](https://doc.rust-lang.org/std/slice/fn.from_raw_parts_mut.html):

> data must be non-null and aligned even for zero-length slices. One reason for this is that enum layout optimizations may rely on references (including slices of any length) being aligned and non-null to distinguish them from other data. You can obtain a pointer that is usable as data for zero-length slices using [NonNull::dangling()](https://doc.rust-lang.org/std/ptr/struct.NonNull.html#method.dangling).

Reply:
Good point. In JVM's docs, it does not mention any enforcement of alignments even for non-empty arrays. I resorted to converting both source and destination pointers to byte pointers and copying with `copy_nonoverlapping`. Also added some overflow checks.